RecoParticleFlow/PFClusterProducer/plugins/alpaka/PFClusterSoAProducerKernel.dev.cc

```diff
@@ -0,0 +1,1634 @@
+              }
+            }
+          }
+        }
+      }
+    }
+  };
+
+  class FastCluster {
+  public:
+    template <typename TAcc, typename = std::enable_if<!std::is_same_v<Device, alpaka::DevCpu>>>
+    ALPAKA_FN_ACC void operator()(const TAcc& acc,
+                                  const reco::PFRecHitHostCollection::ConstView pfRecHits,
+                                  const reco::PFClusterParamsDeviceCollection::ConstView pfClusParams,
+                                  reco::PFClusteringVarsDeviceCollection::View pfClusteringVars,
+                                  reco::PFClusterDeviceCollection::View clusterView,
+                                  reco::PFRecHitFractionDeviceCollection::View fracView) const {
+      const int nRH = pfRecHits.size();
+      int& topoId = alpaka::declareSharedVar<int, __COUNTER__>(acc);
+      int& nRHTopo = alpaka::declareSharedVar<int, __COUNTER__>(acc);
+      int& nSeeds = alpaka::declareSharedVar<int, __COUNTER__>(acc);
+
+      if (once_per_block(acc)) {
+        topoId = alpaka::getIdx<alpaka::Grid, alpaka::Blocks>(acc)[0u];
+        nRHTopo = pfClusteringVars[topoId].topoRHCount();
+        nSeeds = pfClusteringVars[topoId].topoSeedCount();
+      }
+
+      alpaka::syncBlockThreads(acc);  // all threads call sync
+
+      if (topoId < nRH && nRHTopo > 0 && nSeeds > 0) {
+        if (nRHTopo == nSeeds) {
+          // PF cluster is isolated seed. No iterations needed
+          if (once_per_block(acc)) {
+            // Fill PFCluster-level information
+            int rhIdx = pfClusteringVars[pfClusteringVars[topoId].topoSeedOffsets()]
+                            .topoSeedList();  // i is the seed rechit index
+            int seedIdx = pfClusteringVars[rhIdx].rhIdxToSeedIdx();
+            clusterView[seedIdx].energy() = pfRecHits[rhIdx].energy();
+            clusterView[seedIdx].x() = pfRecHits[rhIdx].x();
+            clusterView[seedIdx].y() = pfRecHits[rhIdx].y();
+            clusterView[seedIdx].z() = pfRecHits[rhIdx].z();
+          }
+        } else if (nSeeds == 1) {
+          // Single seed cluster
+          hcalFastCluster_singleSeed(
+              acc, pfClusParams, topoId, nRHTopo, pfRecHits, pfClusteringVars, clusterView, fracView);
+        } else if (nSeeds <= 100 && nRHTopo - nSeeds < threadsPerBlockForClustering) {
+          hcalFastCluster_multiSeedParallel(
+              acc, pfClusParams, topoId, nSeeds, nRHTopo, pfRecHits, pfClusteringVars, clusterView, fracView);
+        } else if (nSeeds <= 400 && nRHTopo - nSeeds <= 1500) {
+          hcalFastCluster_multiSeedIterative(
+              acc, pfClusParams, topoId, nSeeds, nRHTopo, pfRecHits, pfClusteringVars, clusterView, fracView);
+        } else {
+          if (once_per_block(acc))
+            printf("ERROR: Topo cluster %d has %d seeds and %d rechits. Will be processed in next kernel.\n",
```

Comment:
This does not look like an ERROR ?
Also, could you add the `debug` template argument, and make this block `if constexpr (debug)` ?